src/Common/ConcurrencyControl.h

```diff
@@ -0,0 +1,234 @@
+                    std::optional<Waiters::iterator>()};
+        }
+
+        void wait(Waiters::iterator waiter_)
+        {
+            waiter = waiter_;
+        }
+
+        // Grant single slot to allocation, returns true iff more slot(s) are required
+        bool grant()
+        {
+            std::unique_lock lock{mutex};
+            granted++;
+            allocated++;
+            return allocated < limit;
+        }
+
+        // Release one slot and grant it to other allocation if required
+        void release()
+        {
+            parent.release(1);
+            std::unique_lock lock{mutex};
+            released++;
+            assert(released <= allocated);
+        }
+
+        ConcurrencyControl & parent;
+        const SlotCount limit;
+
+        std::mutex mutex; // the following values must be accessed under this mutex
+        SlotCount allocated = 0; // allocated total (including already released)
+        SlotCount granted = 0; // allocated, but not yet acquired
+        SlotCount released = 0;
+
+        Waiters::iterator waiter; // iterator to itself in Waiters list; valid iff allocated < limit
+    };
+
+public:
+    ConcurrencyControl()
+        : cur_waiter(waiters.end())
+    {}
+
+    // WARNING: all Allocation objects MUST be destructed before ConcurrencyControl
+    // NOTE: Recommended way to achieve this is to use `instance()` and do graceful shutdown of queries
+    ~ConcurrencyControl()
+    {
+        assert(waiters.empty());
+    }
+
+    // Allocate at least `min` and at most `max` slots.
+    // If not all `max` slots were successfully allocated, a subscription for later allocation is created
+    // Use Allocation::tryAcquire() to acquire allocated slot, before running a thread.
+    [[nodiscard]] AllocationPtr allocate(SlotCount min, SlotCount max)
+    {
+        std::unique_lock lock{mutex};
+
+        // Acquire as much slots as we can, but not lower than `min`
+        SlotCount limit = std::max(min, max);
+        SlotCount granted = std::max(min, std::min(limit, available(lock)));
+        cur_concurrency += granted;
+
+        // Create allocation and start waiting if more slots are required
+        auto allocation = new Allocation(*this, limit, granted);
+        if (granted < limit)
+            allocation->wait(waiters.insert(cur_waiter, allocation));
+        return AllocationPtr(allocation);
+    }
+
+    void setMaxConcurrency(SlotCount value)
+    {
+        std::unique_lock lock{mutex};
+        max_concurrency = std::max<SlotCount>(1, value); // never allow max_concurrency to be zero
+        schedule(lock);
+    }
+
+    static ConcurrencyControl & instance()
+    {
+        static ConcurrencyControl result;
+        return result;
+    }
+
+private:
+    friend struct Allocation; // for free() and release()
+
+    void free(Allocation * allocation)
+    {
+        std::unique_lock lock{mutex};
+        auto [amount, waiter] = allocation->free();
```

Comment:
Yes, unfortunately it is important to avoid race, `grant()` can happen into the same allocation when we destroy allocation which is still waiting for more slots (this is in fact canceling, that was hard to be done in previous design of upscaling that we discussed with you).

Locking parent's mutex prevents slot scheduling and thus `grant()` to be called. Also in earlier implementation these two classes `ConcurrencyControl` and `Allocation` were more tightly coupled and there was weird `unique_lock lock{parent.mutex}` construction, but @ilejn convinced me to remove it.

Reply:
Ok.
Another option is, probably, add some flag which tells that Allocation is cencelled.
And grant can check that flag and return some special status that it was not successful.
Anyway, a comment will be useful :)